src/python_testing/matter_testing_infrastructure/chip/testing/decorators.py

```diff
@@ -0,0 +1,355 @@
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+"""Decorator utilities for Matter testing infrastructure.
+
+This module provides decorator functions for Matter testing to handle async operations
+and endpoint matching.
+"""
+
+import asyncio
+import logging
+from enum import IntFlag
+from functools import partial
+from typing import TYPE_CHECKING, Callable
+
+import chip.clusters as Clusters
+from chip.clusters import Attribute
+from chip.clusters import ClusterObjects as ClusterObjects
+from chip.testing.global_attribute_ids import GlobalAttributeIds
+from mobly import asserts
+
+if TYPE_CHECKING:
+    from chip.testing.matter_testing import MatterBaseTest
+
+EndpointCheckFunction = Callable[[Clusters.Attribute.AsyncReadTransaction.ReadResponse, int], bool]
+
+
+def get_cluster_from_attribute(attribute: ClusterObjects.ClusterAttributeDescriptor) -> ClusterObjects.Cluster:
+    """Get the Cluster object for an attribute's cluster_id.
+
+    Args:
+        attribute: A ClusterAttributeDescriptor object containing cluster_id
+
+    Returns:
+        ClusterObjects.Cluster: The corresponding Cluster object from ALL_CLUSTERS
+
+    Raises:
+        KeyError: If the attribute's cluster_id is not found in ALL_CLUSTERS dictionary
+    """
+    return ClusterObjects.ALL_CLUSTERS[attribute.cluster_id]
+
+
+def get_cluster_from_command(command: ClusterObjects.ClusterCommand) -> ClusterObjects.Cluster:
+    """Get the Cluster object for a command's cluster_id.
+
+    Args:
+        command: A ClusterCommand object containing cluster_id
+
+    Returns:
+        ClusterObjects.Cluster: The corresponding Cluster object from ALL_CLUSTERS
+
+    Raises:
+        KeyError: If the command's cluster_id is not found in ALL_CLUSTERS dictionary
+    """
+    return ClusterObjects.ALL_CLUSTERS[command.cluster_id]
+
+
+def _has_cluster(wildcard: Clusters.Attribute.AsyncReadTransaction.ReadResponse, endpoint: int, cluster: ClusterObjects.Cluster) -> bool:
+    """Check if a cluster exists on a specific endpoint.
+
+    Args:
+        wildcard: A wildcard read result containing endpoint attributes mapping
+        endpoint: The endpoint ID to check
+        cluster: The Cluster object to look for
+
+    Returns:
+        bool: True if the cluster exists on the endpoint, False otherwise
+            Returns False if endpoint is not found in wildcard attributes
+    """
+    try:
+        return cluster in wildcard.attributes[endpoint]
+    except KeyError:
+        return False
```

Comment:
```suggestion
    return (endpoint in wildcard.attributes) and (cluster in wildcard.attributes[endpoint])
```

maybe clearer and no exception under normal usage.

Or maybe this works (I do not know the type of wildcard attributes, but assuming dict)

```py
    return cluster in wildcard.attributes.get(endpoint, [])
```